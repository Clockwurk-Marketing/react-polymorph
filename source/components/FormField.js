import React, { Component } from 'react';
import PropTypes from 'prop-types';

// import utility functions
import { StringOrElement } from '../utils/props.js';
import composeTheme from '../utils/composeTheme.js';

// import the FormField component's theme API
import { formFieldThemeAPI } from '../themes/API/formField.js';

export default class FormField extends Component {
  state = { error: '' };

  static propTypes = {
    render: PropTypes.func.isRequired,
    skin: PropTypes.func.isRequired,
    label: StringOrElement,
    disabled: PropTypes.bool,
    error: StringOrElement,
    theme: PropTypes.object,
    themeOverrides: PropTypes.object,
    themeAPI: PropTypes.object
  };

  static defaultProps = {
    disabled: false,
    theme: {},
    themeOverrides: {}, // custom css/scss from user that adheres to React Polymorph theme API
    themeAPI: { ...formFieldThemeAPI }
  };

  _setError = error => this.setState({ error });

  render() {
    // destructuring the props here ensures that variable names
    // do not overwrite each other, only pass on the "...rest" of the props

    const {
      skin: FormFieldSkin,
      theme,
      themeOverrides,
      themeAPI,
      error,
      ...rest
    } = this.props;

    const composedTheme = composeTheme(theme, themeOverrides, themeAPI);

    return (
      <FormFieldSkin
        error={error || this.state.error}
        setError={this._setError}
        theme={composedTheme}
        {...rest}
      />
    );
  }
}
