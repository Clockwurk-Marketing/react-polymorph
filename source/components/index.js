// @flow
// exports all components as named exports
export { ThemeProvider } from './ThemeProvider';
export { Autocomplete } from './Autocomplete';
export { Bubble } from './Bubble';
export { Button } from './Button';
export { Checkbox } from './Checkbox';
export { FormField } from './FormField';
export { Input } from './Input';
export { NumericInput } from './NumericInput';
export { Modal } from './Modal';
export { Options } from './Options';
export { Radio } from './Radio';
export { Select } from './Select';
export { TextArea } from './TextArea';
export { Tooltip } from './Tooltip';
