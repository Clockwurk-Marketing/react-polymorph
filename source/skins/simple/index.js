// @flow
// exports all skins as named exports
export { default as AutocompleteSkin } from './AutocompleteSkin';
export { default as BubbleSkin } from './BubbleSkin';
export { default as ButtonSkin } from './ButtonSkin';
export { default as CheckboxSkin } from './CheckboxSkin';
export { default as FormFieldSkin } from './FormFieldSkin';
export { default as InputSkin } from './InputSkin';
export { default as ModalSkin } from './ModalSkin';
export { default as OptionsSkin } from './OptionsSkin';
export { default as RadioSkin } from './RadioSkin';
export { default as SelectSkin } from './SelectSkin';
export { default as SwitchSkin } from './SwitchSkin';
export { default as TextAreaSkin } from './TextAreaSkin';
export { default as TogglerSkin } from './TogglerSkin';
export { default as TooltipSkin } from './TooltipSkin';
