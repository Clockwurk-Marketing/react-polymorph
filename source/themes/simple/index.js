// @flow
// import theme IDENTIFIERS constants
import { IDENTIFIERS } from '../API';

// css modules plugin converts all imports below into plain objects
import SimpleAutocomplete from './SimpleAutocomplete.scss';
import SimpleBubble from './SimpleBubble.scss';
import SimpleButton from './SimpleButton.scss';
import SimpleCheckbox from './SimpleCheckbox.scss';
import SimpleFormField from './SimpleFormField.scss';
import SimpleHeader from './SimpleHeader.scss';
import SimpleInput from './SimpleInput.scss';
import SimpleModal from './SimpleModal.scss';
import SimpleOptions from './SimpleOptions.scss';
import SimpleRadio from './SimpleRadio.scss';
import SimpleSelect from './SimpleSelect.scss';
import SimpleSwitch from './SimpleSwitch.scss';
import SimpleTextArea from './SimpleTextArea.scss';
import SimpleToggler from './SimpleToggler.scss';
import SimpleTooltip from './SimpleTooltip.scss';

// SimpleTheme is a plain object serving as the default export.
// Each key is named after a component and each key's value
// is the component's corresponding theme. The user can
// pass this entire obj directly to ThemeProvider via the "theme" prop
export const SimpleTheme = {
  [IDENTIFIERS.AUTOCOMPLETE]: SimpleAutocomplete,
  [IDENTIFIERS.BUBBLE]: SimpleBubble,
  [IDENTIFIERS.BUTTON]: SimpleButton,
  [IDENTIFIERS.CHECKBOX]: SimpleCheckbox,
  [IDENTIFIERS.FORM_FIELD]: SimpleFormField,
  [IDENTIFIERS.HEADER]: SimpleHeader,
  [IDENTIFIERS.INPUT]: SimpleInput,
  [IDENTIFIERS.MODAL]: SimpleModal,
  [IDENTIFIERS.OPTIONS]: SimpleOptions,
  [IDENTIFIERS.RADIO]: SimpleRadio,
  [IDENTIFIERS.SELECT]: SimpleSelect,
  [IDENTIFIERS.SWITCH]: SimpleSwitch,
  [IDENTIFIERS.TEXT_AREA]: SimpleTextArea,
  [IDENTIFIERS.TOGGLER]: SimpleToggler,
  [IDENTIFIERS.TOOLTIP]: SimpleTooltip
};
