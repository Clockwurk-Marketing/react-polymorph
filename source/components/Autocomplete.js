// @flow
import React, { Component } from "react";
import type { ComponentType } from "react";
import { withTheme } from "../themes/withTheme";

// external libraries
import _ from 'lodash';

// internal utility functions
import { composeTheme, composeFunctions, addThemeId } from "../utils";

import { IDENTIFIERS } from '../themes/API';

type Props = {
  context: {
    theme: Object,
    ROOT_THEME_API: Object
  },
  error: string,
  invalidCharsRegex: RegExp,
  isOpeningUpward: boolean,
  label: string | Element,
  maxSelections: number,
  maxVisibleOptions: number,
  multipleSameSelections: boolean,
  onChange: Function,
  options: Array<any>,
  selectedOptions: Array<any>,
  placeholder: string,
  renderSelections: Function,
  renderOptions: Function,
  skin: ComponentType<any>,
  sortAlphabetically: boolean,
  theme: Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object
};

type State = {
  inputValue: string,
  error: string,
  selectedOptions: Array<any>,
  filteredOptions: Array<any>,
  isOpen: boolean,
  composedTheme: Object
};

class Autocomplete extends Component<Props, State> {
  rootElement: ?Element;
  inputElement: HTMLInputElement;
  suggestionsElement: ?Element;

  static defaultProps = {
    error: null,
    invalidCharsRegex: /[^a-zA-Z0-9]/g, // only allow letters and numbers by default
    isOpeningUpward: false,
    maxVisibleOptions: 10, // max number of visible options
    multipleSameSelections: true, // if true then same word can be selected multiple times
    options: [],
    sortAlphabetically: true, // options are sorted alphabetically by default
    theme: null,
    themeId: IDENTIFIERS.AUTOCOMPLETE,
    themeOverrides: {}
  };

  constructor(props: Props) {
    super(props);

    const {
      context,
      themeId,
      theme,
      themeOverrides,
      sortAlphabetically,
      options,
      selectedOptions
    } = props;

    this.state = {
      inputValue: '',
      error: '',
      selectedOptions: selectedOptions || [],
      filteredOptions:
        sortAlphabetically && options ? options.sort() : options || [],
      isOpen: false,
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      )
    };
  }

  clear = () => this._removeOptions();

  focus = () => this.handleAutocompleteClick();

  openOptions = () => {
    this.setState({ isOpen: true });
  };

  closeOptions = () => {
    this.setState({ isOpen: false });
  };

  handleAutocompleteClick = () => {
    this.inputElement.focus();
    if (!this.state.isOpen) {
      this.openOptions();
    } else {
      this.closeOptions();
    }
  };

  // checks for backspace in order to delete the last selected option
  onKeyDown = (event: SyntheticKeyboardEvent<>) => {
    if (
      event.keyCode === 8 &&
      !event.target.value &&
      this.state.selectedOptions.length
    ) {
      // Remove last selected option
      this.removeOption(this.state.selectedOptions.length - 1, event);
    }
  };

  // onChange handler for input element in AutocompleteSkin
  handleInputChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
    const value = event.target.value;

    // filter out invalid characters
    const filteredValue = this._filterInvalidChars(value);

    // filter options
    const filteredOptions = this._filterOptions(filteredValue);

    // open options, update filteredOptions, and update inputValue
    this.setState({
      isOpen: true,
      inputValue: filteredValue,
      filteredOptions
    });
  };

  // passed to Options onChange handler in AutocompleteSkin
  handleChange = (option: any, event: SyntheticEvent<>) => {
    this.updateSelectedOptions(event, option);
  };

  updateSelectedOptions = (
    event: SyntheticEvent<>,
    selectedOption: any = null
  ) => {
    const canMoreOptionsBeSelected =
      this.state.selectedOptions.length < this.props.maxSelections;
    const areFilteredOptionsAvailable =
      this.state.filteredOptions && this.state.filteredOptions.length > 0;

    if (
      !this.props.maxSelections ||
      (canMoreOptionsBeSelected && areFilteredOptionsAvailable)
    ) {
      if (!selectedOption) return;
      const option = selectedOption.trim();
      const optionCanBeSelected =
        (this.state.selectedOptions.indexOf(option) < 0 &&
          !this.props.multipleSameSelections) ||
        this.props.multipleSameSelections;

      if (option && optionCanBeSelected && this.state.isOpen) {
        const selectedOptions = _.concat(this.state.selectedOptions, option);
        this.selectionChanged(selectedOptions, event);
        this.setState({ selectedOptions, isOpen: false });
      }
    }

    this.setState({ inputValue: '' });
  };

  removeOption = (index: number, event: SyntheticEvent<>) => {
    const selectedOptions = this.state.selectedOptions;
    _.pullAt(selectedOptions, index);
    this.selectionChanged(selectedOptions, event);
    this.setState({ selectedOptions });
  };

  selectionChanged = (
    selectedOptions: Array<any>,
    event: SyntheticEvent<any>
  ) => {
    if (this.props.onChange) this.props.onChange(selectedOptions, event);
  };

  // returns an object containing props, theme, and method handlers
  // associated with rendering this.state.selectedOptions, the user can call
  // this in the body of the renderSelections function
  getSelectionProps = ({
    removeSelection
  }: { removeSelection: Function } = {}) => {
    const { themeId } = this.props;
    const { inputValue, isOpen, selectedOptions, composedTheme } = this.state;
    return {
      inputValue,
      isOpen,
      selectedOptions,
      theme: composedTheme[themeId],
      removeSelection: (index: number, event: SyntheticEvent<>) =>
        // the user's custom removeSelection event handler is composed with
        // the internal functionality of Autocomplete (this.removeOption)
        composeFunctions(removeSelection, this.removeOption)(index, event)
    };
  };

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const {
      skin: AutocompleteSkin,
      theme,
      themeOverrides,
      onChange,
      error,
      ...rest
    } = this.props;

    return (
      <AutocompleteSkin
        inputValue={this.state.inputValue}
        selectedOptions={this.state.selectedOptions}
        filteredOptions={this.state.filteredOptions}
        isOpen={this.state.isOpen}
        theme={this.state.composedTheme}
        handleInputChange={this.handleInputChange}
        error={error || this.state.error}
        rootRef={el => (this.rootElement = el)}
        inputRef={el => (this.inputElement = el)}
        suggestionsRef={el => (this.suggestionsElement = el)}
        handleChange={this.handleChange}
        closeOptions={this.closeOptions}
        handleAutocompleteClick={this.handleAutocompleteClick}
        removeOption={this.removeOption}
        onKeyDown={this.onKeyDown}
        getSelectionProps={this.getSelectionProps}
        {...rest}
      />
    );
  }

  // ======== PRIVATE METHOD ==========

  _removeOptions = () => {
    const { onChange } = this.props;
    onChange ? onChange([]) : null;
    this.setState({ selectedOptions: [], inputValue: '' });
  };

  _filterOptions = (value: string) => {
    let filteredOptions = [];

    if (value !== '') {
      _.some(this.props.options, (option) => {
        if (_.startsWith(option, value)) {
          filteredOptions.push(option);
        }
      });
    } else {
      filteredOptions = this.props.options;
    }

    return filteredOptions;
  };

  _filterInvalidChars = (value: string) => {
    let filteredValue = '';

    if (this.props.invalidCharsRegex.test(value)) {
      filteredValue = value.replace(this.props.invalidCharsRegex, '');
    } else {
      filteredValue = value;
    }

    return filteredValue;
  };
}

export default withTheme(Autocomplete);
