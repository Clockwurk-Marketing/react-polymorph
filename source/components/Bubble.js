import React, { Component } from 'react';
import PropTypes from 'prop-types';

// import utility functions
import composeTheme from '../utils/composeTheme.js';

// import the Bubble component's theme API
import { BUBBLE_THEME_API } from '../themes/API';

class Bubble extends Component {
  static propTypes = {
    isOpeningUpwards: PropTypes.bool,
    isTransparent: PropTypes.bool,
    skin: PropTypes.func.isRequired,
    theme: PropTypes.object,
    themeOverrides: PropTypes.object,
    themeAPI: PropTypes.object
  };

  static defaultProps = {
    isTransparent: true,
    theme: {},
    themeOverrides: {},
    themeAPI: { ...BUBBLE_THEME_API }
  };

  constructor(props, context) {
    super(props);

    const { themeOverrides, themeAPI } = props;

    const theme =
      context && context.theme && context.theme.bubble
        ? context.theme.bubble
        : props.theme;

    // if themeOverrides isn't provided, composeTheme returns theme immediately
    this.state = {
      composedTheme: composeTheme(theme, themeOverrides, themeAPI)
    };
  }

  render() {
    const {
      skin: BubbleSkin,
      theme,
      themeOverrides,
      themeAPI,
      ...rest
    } = this.props;

    return <BubbleSkin theme={this.state.composedTheme} {...rest} />;
  }
}

Bubble.contextTypes = {
  theme: PropTypes.object
};

export default Bubble;
