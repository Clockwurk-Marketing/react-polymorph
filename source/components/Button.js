// @flow
import React, { Component } from 'react';
import type { ComponentType } from 'react';
import { withTheme } from '../themes/withTheme';

// internal utility functions
import { composeTheme, addThemeId } from '../utils';

// import constants
import { IDENTIFIERS } from '../themes/API';

type Props = {
  className: string,
  context: {
    theme: Object,
    ROOT_THEME_API: Object
  },
  disabled: boolean,
  label: string | Element,
  onClick: Function,
  skin: ComponentType<any>,
  theme: Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object // custom css/scss from user that adheres to component's theme API
};

type State = {
  composedTheme: Object
};

class Button extends Component<Props, State> {
  static defaultProps = {
    disabled: false,
    theme: null,
    themeId: IDENTIFIERS.BUTTON,
    themeOverrides: {}
  };

  constructor(props: Props) {
    super(props);

    const { context, themeId, theme, themeOverrides } = props;

    this.state = {
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      )
    };
  }

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const {
      skin: ButtonSkin,
      theme,
      themeOverrides,
      context,
      ...rest
    } = this.props;

    return <ButtonSkin theme={this.state.composedTheme} {...rest} />;
  }
}

export default withTheme(Button);
