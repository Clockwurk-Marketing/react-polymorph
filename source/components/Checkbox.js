// @flow
import React, { Component } from 'react';
import type { ComponentType } from 'react';
import { withTheme } from '../themes/withTheme';

// import utility functions
import { composeTheme, addThemeId } from '../utils';

// import constants
import { IDENTIFIERS } from '../themes/API';

type Props = {
  checked: boolean,
  context: {
    theme: Object,
    ROOT_THEME_API: Object
  },
  disabled: boolean,
  label: string | Element,
  onChange: Function,
  onBlur: Function,
  onFocus: Function,
  skin: ComponentType<any>,
  theme: Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object
};

type State = {
  composedTheme: Object
};

class Checkbox extends Component<Props, State> {
  static defaultProps = {
    checked: false,
    disabled: false,
    theme: null,
    themeId: IDENTIFIERS.CHECKBOX,
    themeOverrides: {}
  };

  constructor(props: Props) {
    super(props);

    const { context, themeId, theme, themeOverrides } = props;

    this.state = {
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      )
    };
  }

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const { skin: CheckboxSkin, theme, themeOverrides, ...rest } = this.props;

    return <CheckboxSkin theme={this.state.composedTheme} {...rest} />;
  }
}

export default withTheme(Checkbox);
