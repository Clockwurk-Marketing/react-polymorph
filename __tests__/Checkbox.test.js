import React from 'react';
import renderer from 'react-test-renderer';

import { Checkbox } from '../source/components';
import { CheckboxSkin } from '../source/skins/simple';
import CheckboxTheme from '../source/themes/simple/SimpleCheckbox.scss';

const SimpleTheme = { checkbox: CheckboxTheme };

test('Checkbox renders to the DOM', () => {
  const component = renderer.create(
    <Checkbox theme={SimpleTheme} skin={CheckboxSkin} />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Checkbox renders with a label', () => {
  const component = renderer.create(
    <Checkbox label="check here" theme={SimpleTheme} skin={CheckboxSkin} />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Checkbox is disabled', () => {
  const component = renderer.create(
    <Checkbox disabled theme={SimpleTheme} skin={CheckboxSkin} />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Checkbox is checked', () => {
  const component = renderer.create(
    <Checkbox checked theme={SimpleTheme} skin={CheckboxSkin} />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
