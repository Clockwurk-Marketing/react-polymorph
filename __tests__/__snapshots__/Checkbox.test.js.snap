// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox is checked 1`] = `
<div
  aria-hidden={true}
  className="root checked"
  onClick={[Function]}
  role="presentation"
>
  <input
    checked={true}
    className="input"
    disabled={false}
    type="checkbox"
  />
  <div
    className="check checked"
  />
</div>
`;

exports[`Checkbox is disabled 1`] = `
<div
  aria-hidden={true}
  className="root disabled"
  onClick={[Function]}
  role="presentation"
>
  <input
    checked={false}
    className="input"
    disabled={true}
    type="checkbox"
  />
  <div
    className="check"
  />
</div>
`;

exports[`Checkbox renders to the DOM 1`] = `
<div
  aria-hidden={true}
  className="root"
  onClick={[Function]}
  role="presentation"
>
  <input
    checked={false}
    className="input"
    disabled={false}
    type="checkbox"
  />
  <div
    className="check"
  />
</div>
`;

exports[`Checkbox renders with a label 1`] = `
<div
  aria-hidden={true}
  className="root"
  onClick={[Function]}
  role="presentation"
>
  <input
    checked={false}
    className="input"
    disabled={false}
    label="check here"
    type="checkbox"
  />
  <div
    className="check"
  />
  <label
    className="label"
  >
    check here
  </label>
</div>
`;
